### PR TITLE
Make KernelBoolSlice memory safe, and Send+Sync

### PR DESCRIPTION
**Pull request by scovich**
Initial work on https://github.com/delta-incubator/delta-kernel-rs/pull/196 exposed a lurking design issue with the FFI `KernelBoolSlice`. By using raw pointers, it cannot be Send+Sync, but also memory lifetimes become very hard to reason about.

The fix is to pass it to the engine by value, with the contract that engine must free it. Also, `unsafe impl` Send+Sync on the struct, with explanation of the safety requirements.

This is a temporary fix, since we'd prefer kernel didn't allocate boolean slices at all if possible, and definitely shouldn't pass them as owned into visitor methods that anyway just drop it.

### CODE REVIEW

File: ffi/src/lib.rs [hunk 1]
```diff
@@ -115,40 +115,85 @@ impl TryFromStringSlice for String {
...
-        let ptr = Box::into_raw(boxed).cast();
-        KernelBoolSlice { ptr, len }
-    }
-}
+        /// Converts this slice back into a `Vec<bool>`.
+        ///
+        /// # Safety
+        ///
+        /// The slice must have been originally created `From<Vec<bool>>`, and must not have been
+        /// already been consumed by a previous call tothis method.
```
Comment (nicklan) [inline ffi/src/lib.rs:144]:
```suggestion
        /// already been consumed by a previous call to this method.
```

File: ffi/src/lib.rs [hunk 2]
```diff
@@ -115,40 +115,85 @@ impl TryFromStringSlice for String {
...
 
-trait FromBoolSlice {
-    unsafe fn from_slice(slice: KernelBoolSlice) -> Self;
-}
+        /// Converts this slice back into a `Vec<bool>`.
+        ///
+        /// # Safety
+        ///
+        /// The slice must have been originally created `From<Vec<bool>>`, and must not have been
+        /// already been consumed by a previous call tothis method.
```
Comment (nicklan) [inline ffi/src/lib.rs:158]:
```suggestion
        /// already been consumed by a previous call to this method.
```

**Review by nicklan — APPROVED** [review]
lgtm (with two minor comment fixes)

File: ffi/src/scan.rs [hunk 3]
```diff
@@ -330,17 +329,16 @@ struct ContextWrapper {
 #[no_mangle]
 pub unsafe extern "C" fn visit_scan_data(
     data: *mut EngineDataHandle,
-    selection_vector: KernelBoolSlice,
+    selection_vec: KernelBoolSlice,
     engine_context: NullableCvoid,
     callback: CScanCallback,
 ) {
-    let selection_vec = Vec::<bool>::from_slice(selection_vector);
+    let selection_vec = unsafe { selection_vec.as_ref() };
```
Comment (zachschuermann) [inline ffi/src/scan.rs:336]: nit: safety comment?

Comment (zachschuermann) [reply ffi/src/scan.rs:336]: though i suppose this is in unsafe extern anyways...

**Review by zachschuermann — APPROVED** [review]
lgtm

File: ffi/src/scan.rs (hunk 3 above)
Comment (scovich) [reply ffi/src/scan.rs:336]: Yeah, `visit_scan_data` safety comment eventually needs to be updated, but prob not as part of this PR

**Pull request merged**